### PR TITLE
Fix post-processing crash when disabling and re-enabling stages

### LINKED ISSUE #9204: Postprocessing is initially disabled and will fail if enabled again
Click the example's "enable" checkbox.

Sandcastle example:

[https://sandcastle.cesium.com/#c=fVVRb+I4EP4rFk+hUAMt190rUN0VWN3qdrfVstp7iYScxIBbx45sh5au+t9vxgmQQO5aqfWMZ76Z+cYz2TJDtoK/cEMmRPEXMuVW5Cn96XVB2Iq9PNXKMaG4CVtd8itUhNiNzmXypxIpc/yWOJPzbqje26NQhWoLqJm2wgmtALfEnDLj4MTUNV0Znc742nBug8vB1TXtfxgObwa/d8lwSPu/9a8/9G8QCoFyIwEjbFHaW7A0k3zGHOulOuHS9gror0yVpyUc6VpGYQu9fRHUGRY/82SunHA7gCrVHGXBLWVJEviaFEuhFIjXRWlfwO3h5NU+8G1BAgFjcfR4rzOwMmydQpTFhiXcLHRuYk4maBi2ciVW2qTE+orM1YzEWmrzg7+63PARCQGkRTqFMYDthFqTLY+vyHbpCqOp1iYRCui3J/axVtYRoRz5e/792/zL8p/Psx9/QeWDm1NkLRKSQmcDPLXrl7/qIpbrM7COZwAWv6XLTLxy+Z0BOaTnFchtpo2jby+j/3CHvKDzTD5qi81oKIdcIstB01WXfKR9cuFzaDdHuCZsy4F6/pPJnGMI0AV92m+yhyaQAJkSYNgfwb9xjTTQdDrtc78GbmpwTwXc0xncUyPc/0DiT62gzoSUvFzNguqr6dao7RR9uvCkB6JLntptatbRqDnM+7m6QVVLpDchK6mZC2qv7KJWcBPna7n8BChTTL1ozzCoAnfJ4LxZZTIwXMXrhpl0j0bDerLHkbYxV5xWrhYOQIsJ9wN+XHCPJ0ZBOdD1ob1tHOJi2NuhOs46JvAVNwMk45G4YpHkCQAwaf1q9LZl9Gel42edu2I3BQfv/c5zWsuIITuJjnOMT9fczSXH4/3uM5TTKm3CFjqd4rIsk7t7oWBs1vaI390jo49/rBgN9x68nWMR7aIGsSJHV7ph9uFFAWcZN24XoFO7vd+DJ/ELJSGY9UNkudkiG9U8vPvBzOaRjY2IeJBnCcx5pTuYKbANBCJ/q1zF/qtyZheUuVS6T8smAI33UDZnqlJOeYfwgNwA5/sVqlbXP9qxdTvJ7/YZ/yFSXHO4+gP4MDkOaxz3Uy/K4VvjaLxPHDx7VddxIrZEJJOGLyuJJbMWbla5lAvxxsPW3bgH9meuMHXY2AcYGsl2aLYZ3H0plJTScQ/EZs/DozlUMnbIw0FERaSTXUWBKlOTUZPczQsGxz04n9+OhcpyR9wu41jshsfPkX6FKoFodhnB09yrcUrKbviKT+BAUY0OYi0/kCsF7KvGX5SLv/8C](https://sandcastle.cesium.com/#c=fVVRb+I4EP4rFk+hUAMt190rUN0VWN3qdrfVstp7iYScxIBbx45sh5au+t9vxgmQQO5aqfWMZ76Z+cYz2TJDtoK/cEMmRPEXMuVW5Cn96XVB2Iq9PNXKMaG4CVtd8itUhNiNzmXypxIpc/yWOJPzbqje26NQhWoLqJm2wgmtALfEnDLj4MTUNV0Znc742nBug8vB1TXtfxgObwa/d8lwSPu/9a8/9G8QCoFyIwEjbFHaW7A0k3zGHOulOuHS9gror0yVpyUc6VpGYQu9fRHUGRY/82SunHA7gCrVHGXBLWVJEviaFEuhFIjXRWlfwO3h5NU+8G1BAgFjcfR4rzOwMmydQpTFhiXcLHRuYk4maBi2ciVW2qTE+orM1YzEWmrzg7+63PARCQGkRTqFMYDthFqTLY+vyHbpCqOp1iYRCui3J/axVtYRoRz5e/792/zL8p/Psx9/QeWDm1NkLRKSQmcDPLXrl7/qIpbrM7COZwAWv6XLTLxy+Z0BOaTnFchtpo2jby+j/3CHvKDzTD5qi81oKIdcIstB01WXfKR9cuFzaDdHuCZsy4F6/pPJnGMI0AV92m+yhyaQAJkSYNgfwb9xjTTQdDrtc78GbmpwTwXc0xncUyPc/0DiT62gzoSUvFzNguqr6dao7RR9uvCkB6JLntptatbRqDnM+7m6QVVLpDchK6mZC2qv7KJWcBPna7n8BChTTL1ozzCoAnfJ4LxZZTIwXMXrhpl0j0bDerLHkbYxV5xWrhYOQIsJ9wN+XHCPJ0ZBOdD1ob1tHOJi2NuhOs46JvAVNwMk45G4YpHkCQAwaf1q9LZl9Gel42edu2I3BQfv/c5zWsuIITuJjnOMT9fczSXH4/3uM5TTKm3CFjqd4rIsk7t7oWBs1vaI390jo49/rBgN9x68nWMR7aIGsSJHV7ph9uFFAWcZN24XoFO7vd+DJ/ELJSGY9UNkudkiG9U8vPvBzOaRjY2IeJBnCcx5pTuYKbANBCJ/q1zF/qtyZheUuVS6T8smAI33UDZnqlJOeYfwgNwA5/sVqlbXP9qxdTvJ7/YZ/yFSXHO4+gP4MDkOaxz3Uy/K4VvjaLxPHDx7VddxIrZEJJOGLyuJJbMWbla5lAvxxsPW3bgH9meuMHXY2AcYGsl2aLYZ3H0plJTScQ/EZs/DozlUMnbIw0FERaSTXUWBKlOTUZPczQsGxz04n9+OhcpyR9wu41jshsfPkX6FKoFodhnB09yrcUrKbviKT+BAUY0OYi0/kCsF7KvGX5SLv/8C)

However, if the enable property is true initially, it's fine.

```diff
var viewer = new Cesium.Viewer("cesiumContainer", {
  shouldAnimate: true,
});

var position = Cesium.Cartesian3.fromDegrees(-123.0744619, 44.0503706);
var url = "../SampleData/models/CesiumMan/Cesium_Man.glb";
viewer.trackedEntity = viewer.entities.add({
  name: url,
  position: position,
  model: {
    uri: url,
  },
});

var fragmentShaderSource =
  "uniform sampler2D colorTexture; \n" +
  "varying vec2 v_textureCoordinates; \n" +
  "const int KERNEL_WIDTH = 16; \n" +
  "void main(void) \n" +
  "{ \n" +
  "    vec2 step = czm_pixelRatio / czm_viewport.zw; \n" +
  "    vec2 integralPos = v_textureCoordinates - mod(v_textureCoordinates, 8.0 * step); \n" +
  "    vec3 averageValue = vec3(0.0); \n" +
  "    for (int i = 0; i < KERNEL_WIDTH; i++) \n" +
  "    { \n" +
  "        for (int j = 0; j < KERNEL_WIDTH; j++) \n" +
  "        { \n" +
  "            averageValue += texture2D(colorTexture, integralPos + step * vec2(i, j)).rgb; \n" +
  "        } \n" +
  "    } \n" +
  "    averageValue /= float(KERNEL_WIDTH * KERNEL_WIDTH); \n" +
  "    gl_FragColor = vec4(averageValue, 1.0); \n" +
  "} \n";
const postProcess = viewer.scene.postProcessStages.add(
  new Cesium.PostProcessStage({
    fragmentShader: fragmentShaderSource,
  })
);

var viewModel = {
-  enabled: false,
+  enabled: true,
};

Cesium.knockout.track(viewModel);
var toolbar = document.getElementById("toolbar");
Cesium.knockout.applyBindings(viewModel, toolbar);
for (var name in viewModel) {
  if (viewModel.hasOwnProperty(name)) {
    Cesium.knockout
      .getObservable(viewModel, name)
      .subscribe(updatePostProcess);
  }
}

function updatePostProcess() {
  postProcess.enabled = Boolean(viewModel.enabled);
}
updatePostProcess();
```

Browser: Chrome 86

Operating System: Windows 10

BTW, the fragment shader is from [https://sandcastle.cesium.com/?src=Custom%20Post%20Process.html&label=Post%20Processing](https://sandcastle.cesium.com/?src=Custom%20Post%20Process.html&label=Post%20Processing)

### LINKED ISSUE #9528: Crash in Post Process Sandcastle
1. Open https://sandcastle.cesium.com/index.html?src=Post%20Processing.html&label=All
2. Uncheck Silhouette
3. Check Night Vision

This will crash every time in all browsers.

### PR DESCRIPTION
**Pull request by j9liu**
Fixes #9204, also fixes #9528.

The crashes were happening because the texture cache wouldn't create framebuffers if no stages were marked as active, despite the fact that they are `enabled`. It will now create framebuffers if at least one stage is enabled.

I ran the unit tests and looked through all of the post-processing Sandcastles, and I don't believe this change introduced any other bugs.

cc: @lilleyse @ebogo1

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @j9liu!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by lilleyse** [comment]
I can confirm the fix works but it might be worth looking into the alternate approach I mentioned in https://github.com/CesiumGS/cesium/issues/9204#issuecomment-870012795. I was curious what would happen so I hacked together a quick version of it and was surprised to see that it fixed the original issue and a totally separate issue https://github.com/CesiumGS/cesium/issues/8345 with the full screen flash. I didn't really dive deep into why it fixed that but maybe it has something to do with the framebuffer and active stages being slightly out of sync with each other, like a one frame delay?

**Comment by lilleyse** [comment]
Also, please add a test to `PostProcessStageCollectionSpec.js`

File: CHANGES.md [hunk 1]
```diff
@@ -24,6 +24,7 @@
 - Fixed misleading documentation in `Matrix4.inverse` and `Matrix4.inverseTransformation` that used "affine transformation" instead of "rotation and translation" specifically. [#9608](https://github.com/CesiumGS/cesium/pull/9608)
 - Fixed a regression where external images in glTF models were not being loaded with `preferImageBitmap`, which caused them to decode on the main thread and cause frame rate stuttering. [#9627](https://github.com/CesiumGS/cesium/pull/9627)
 - Fixed misleading "else" case condition for `color` and `show` in `Cesium3DTileStyle`. A default `color` value is used if no `color` conditions are given. The default value for `show`, `true`, is used if no `show` conditions are given. [#9633](https://github.com/CesiumGS/cesium/pull/9633)
+- Fixed a crash that occurred after disabling and re-enabling a post-processing shader. [#9649](https://github.com/CesiumGS/cesium/pull/9649)
```
Comment (lilleyse) [inline CHANGES.md:27]:
```suggestion
- Fixed a crash that occurred after disabling and re-enabling a post-processing stage. [#9649](https://github.com/CesiumGS/cesium/pull/9649)
```

**Comment by j9liu** [comment]
> I can confirm the fix works but it might be worth looking into the alternate approach I mentioned in [#9204 (comment)](https://github.com/CesiumGS/cesium/issues/9204#issuecomment-870012795). I was curious what would happen so I hacked together a quick version of it and was surprised to see that it fixed the original issue and a totally separate issue #8345 with the full screen flash. I didn't really dive deep into why it fixed that but maybe it has something to do with the framebuffer and active stages being slightly out of sync with each other, like a one frame delay?

I tried to implement that fix, but the extra `update` call actually causes some of the unit tests to fail. I'll start looking into the full screen flash issue to find a specific fix for that.

**Comment by j9liu** [comment]
> maybe it has something to do with the framebuffer and active stages being slightly out of sync with each other, like a one frame delay?

It turned out to be something like that: `PostProcessStageCollection.prototype.execute` only executes active stages, so it skipped over the ones that were `enabled` but not active. The yellow screen flash was happening in the Sandcastle because it would execute the Silhouette stage, but not the next enabled one, so the pure yellow was the result of an incomplete pass. I added a call to check if the stages could be marked active before it tried to `execute` the pipeline, this should fix #8345.

**Comment by lilleyse** [comment]
> I tried to implement that fix, but the extra `update` call actually causes some of the unit tests to fail. I'll start looking into the full screen flash issue to find a specific fix for that.

Hm I'm not seeing those errors so maybe there was a difference in how we implemented it. Here's the code I added to the end of `PostProcessStageCollection.prototype.update`:

```js
  count = 0;
  for (i = 0; i < length; ++i) {
    stage = stages[i];
    if (stage.ready && stage.enabled && stage._isSupported(context)) {
      count++;
    }
  }

  activeStagesChanged = count !== activeStages.length;
  if (activeStagesChanged) {
    this.update(context, useLogDepth, useHdr);
  }
```

Ultimately we arrived at similar fixes because both fixes call `update` a second time to update the active stages, just in different locations. See if unit tests pass for you with that code, I prefer it mainly because it's less changes overall and more self contained.

**Comment by j9liu** [comment]
@lilleyse - that fix worked. I stripped my changes and added your fix instead. I also got rid of the `lastLength` variable (the one I renamed `lastActiveLength`) because it was being used for the same purpose as `activeStagesChanged`.

Do you want me to do anything about the tonemapping issue I mentioned over Slack?
> While writing the tests for PostProcessStageCollection I realized that the tonemapping aspect of it never resets after each test, so when I put some new tests after the tonemapping tests, they were still being incorrectly tonemapped. But it doesn't look like there's an option to disable tonemapping in general -- can I add one?
3:37
I'm not sure the intended functionality -- it looks like PostProcessStageCollection defaults to this ToneMapper.ACES setting, not sure if that's supposed to be "normal" coloring because even when I try to use this setting the test is still being tonemapped. Is it okay for there to be a lack of a tonemapper or is that wrong?

**Comment by lilleyse** [comment]
Tonemapping is only applied when `scene.highDynamicRange = true` so as long as the tonemapping tests set this back to `false` the other tests should run fine.

I think the problem is that those tests are not removing the primitive they add to the scene so it interferes with subsequent tests. If you add `scene.primitives.removeAll();` in `afterEach` it should fix any problems in the future.

**Comment by j9liu** [comment]
@lilleyse - done! I temporarily moved my tests after the tonemapping ones just to be sure the `scene` was resetting and they passed.

**Comment by lilleyse** [comment]
Thanks @j9liu!

**Pull request merged**